src-electron/generator/matter/darwin/Framework/CHIP/templates/helper.js

```diff
@@ -739,42 +739,147 @@ async function availability(clusterName, options) {
+  }
+
+  const data = fetchAvailabilityData(global);
 
   let introducedRelease = findReleaseForPath(
     data,
     ['introduced', ...path],
     options
   );
   if (introducedRelease === undefined) {
-    return false;
+    return undefined;
+  }
+
+  let referenceIndex = data.indexOf(referenceRelease);
+  let introducedIndex = data.indexOf(introducedRelease);
+  if (introducedIndex < referenceIndex) {
+    return -1;
+  }
+
+  if (introducedIndex > referenceIndex) {
+    return 1;
   }
 
+  return 0;
+}
+
+function wasIntroducedBeforeRelease(releaseName, clusterName, options) {
+  const data = fetchAvailabilityData(this.global);
+  const path = makeAvailabilityPath(clusterName, options);
+
   let referenceRelease = findReleaseByName(data, releaseName);
   if (referenceRelease === undefined) {
     throw new Error(`Invalid release name: ${releaseName}`);
   }
 
-  return data.indexOf(introducedRelease) < data.indexOf(referenceRelease);
+  let comparisonStatus = compareIntroductionToReferenceRelease(
+    this.global,
+    makeAvailabilityPath(clusterName, options),
+    options, referenceRelease
+  );
+  if (comparisonStatus === undefined) {
+    // Not introduced yet, so not introduced before anything in particular.
+    return false;
+  }
+
+  return comparisonStatus == -1;
 }
 
-function wasRemoved(cluster, options) {
-  const data = fetchAvailabilityData(this.global);
-  const path = makeAvailabilityPath(cluster, options);
+/**
+ * Utility for wasRemoved and findProvisionalRelease.  Finds a release that
+ * mentions the given path or some ancestor of it in the given section.  Returns
+ * the release and the path that ended up being found, or undefined if nothing
+ * was found.
+ */
+function findReleaseForPathOrAncestorAndSection(global, cluster, options, section) {
+  const data = fetchAvailabilityData(global);
+  let path = makeAvailabilityPath(cluster, options);
 
-  let removedRelease = undefined;
-  let removalPath = [...path];
-  while (removedRelease === undefined && removalPath !== undefined) {
-    removedRelease = findReleaseForPath(
+  while (path !== undefined) {
+    let foundRelease = findReleaseForPath(
       data,
-      ['removed', ...removalPath],
+      [section, ...path],
       options
     );
-    removalPath = findPathToContainer(removalPath);
+    if (foundRelease !== undefined) {
+      return { release: foundRelease, path: path };
+    }
+    path = findPathToContainer(path);
+  }
+  return undefined;
+}
+
+function wasRemoved(cluster, options) {
+  return findReleaseForPathOrAncestorAndSection(this.global, cluster, options, 'removed') !== undefined;
+}
+
+function pathsEqual(path1, path2) {
+  if (path1.length != path2.length) {
+    return false;
+  }
+
+  for (let i = 0; i < path1.length; ++i) {
+    if (path1[i] != path2[i]) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function isSupported(cluster, options) {
+  if (wasRemoved.call(this, cluster, options)) {
```

Comment:
Is it necessary to call this with `.call` ?

Reply:
Yes, because it uses `this` to get `this.global`.  I could rejigger it to call some helper function that takes the global as an arg, and call it from there and here, but this was simpler for now.